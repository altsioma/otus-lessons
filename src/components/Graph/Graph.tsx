import React from "react";
import { SVGPath } from "../SVGPath/SVGPath";
import {
  GraphDataType,
  ColorSetType,
  AreaType,
  AnimationStateType,
  AnimationSpeedType,
} from "../../types";
import { Axis } from "../Axis/Axis";
import { GraphBody } from "../GraphBody/GraphBody";
import { withAnimate } from "../../hoc/withAnimate";

interface IGraphProps {
  data: GraphDataType;
  options: {
    area: AreaType;
    color: ColorSetType;
    multiplier?: number;
  };
  speed: AnimationSpeedType;
  playState: AnimationStateType;
  className?: string;
  onAnimationStateChange: (buttonState: AnimationStateType) => void;
}

const Graphic: React.FC<IGraphProps> = ({ ...props }) => {
  const { multiplier = 1, color } = props.options;
  const [x, y] = props.data[0];
  const firstGraphPoint = { x, y };
  const AnimateGraph = withAnimate(SVGPath);

  return (
    <AnimateGraph
      className={props.className}
      color={color}
      strokeWidth={3}
      coords={{
        offset: firstGraphPoint,
        multiplier: multiplier,
        data: props.data,
      }}
      onAnimationStateChange={props.onAnimationStateChange}
      speed={props.speed}
      playState={props.playState}
    />
  );
};

export const Graph: React.FC<IGraphProps> = ({ ...props }) => {
  const { area } = props.options;
  return (
    <svg width={area.width} height={area.height}>
      <Axis x={area.width} y={area.height} />
      <GraphBody area={area} />
      {props.data.length > 0 && <Graphic {...props} className="graphic" />}
    </svg>
  );
};
