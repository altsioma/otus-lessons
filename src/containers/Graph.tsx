import React from "react";
import { Path } from "../components/Path";
import { GraphData } from "../types/dataset";

type Props = {
  data: GraphData
}

const prepareData = (data: GraphData): string => {
  const x = 20;
  const y = 500;
  const offset = [`M ${x} ${y}`];
  const collection = data.map(section => {
    const xNext = x + section[0] * 20;
    const yNext = y + section[1] * 20;
    return `L ${xNext} ${yNext}`;
  })
  return offset.concat(collection).join(' ');
}

export const Graph: React.FC<Props> = ({ data }) => {
  const coords = prepareData(data);
  return (
    <svg width={800} height={600} >
      <Path coords={coords} color="red" strokeWidth={2} />
    </svg>
  )
};
