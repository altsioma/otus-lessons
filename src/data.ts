import { InitialConfigType, GraphDataType } from "./types/";

export const InitialConfig: InitialConfigType = Object.freeze({
  color: "blue",
  graph: [
    [1, 3],
    [2, 5],
    [3, 2],
    [4, 16],
    [18, 5],
  ],
  colorSet: ["red", "green"],
  area: {
    width: 600,
    height: 300,
  },
});

export const stableGraph: GraphDataType = [
  [1, 3],
  [2, 5],
  [3, 2],
  [4, 16],
  [18, 5],
];
