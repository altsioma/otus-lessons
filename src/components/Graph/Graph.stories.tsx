import React from "react";
import { withKnobs, optionsKnob } from "@storybook/addon-knobs";
import { Graph } from "./Graph";
import { ColorSetType } from "@/types";
import { InitialConfig } from "@/data";

export default {
  title: "GraphMakerComponents",
  decorators: [withKnobs],
};

const label = "Colors";
const options: Record<ColorSetType, string> = {
  red: "red",
  green: "green",
  blue: "blue",
  violet: "violet",
};
const defaultValue = "red";

const value = optionsKnob(
  label,
  options,
  defaultValue,
  {
    display: "inline-radio",
  },
  "Colors"
);

const graphGroupId = "Graph Parameters";

export const GraphBody: React.FC<{}> = () => (
  <Graph
    options={{ area: { width: 600, height: 300 }, color: value, multiplier: 1 }}
    data={InitialConfig.graph}
    playState={"stopped"}
  />
);
