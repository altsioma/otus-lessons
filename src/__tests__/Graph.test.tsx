import React from "react";
import { shallow } from "enzyme";

import { Graph } from "../components/Graph/Graph";
import { stableGraph } from "../data";

describe("Graph render check", () => {

  it("Graph svg render with stable data", () => {
    const wrapper = shallow(<Graph      
      options={{area: {width: 300, height: 300}, color: "red"}}
      data={stableGraph}
      speed={1}
      playState={"stopped"}
      onAnimationStateChange={jest.fn()}
    />)
    expect(
      wrapper.html()
    ).toMatchSnapshot();
  });
  it('Graph with init param data=[[20,40]] with offset x = 0, y = 0 has coords="M 0 0 L 20 40"', () => {
    const wrapper = shallow(<Graph
      options={{area: {width: 300, height: 300}, color: "violet"}}
      data={[[20, 40]]}
      speed={1}
      playState={"stopped"}
      onAnimationStateChange={jest.fn()}
    />)
    expect(
      wrapper.children().find('.graphic').render().attr("d")
    ).toBe("M 20 40 L 20 40")
  })
  it('Graph with init param data=[[20,40],[30,40]] with offset x = 0, y = 0 has coords="M 0 0 L 20 40 L 30 40"', () => {
    const wrapper = shallow(<Graph      
      options={{area: {width: 300, height: 300}, color: "blue"}}
      speed={1}
      playState={"stopped"}
      onAnimationStateChange={jest.fn()}
      data={[[20, 40], [30, 40]]}
    />)
    expect(
      wrapper.children().find('.graphic').render().attr("d")
    ).toBe("M 20 40 L 20 40 L 30 40")
  })

});
