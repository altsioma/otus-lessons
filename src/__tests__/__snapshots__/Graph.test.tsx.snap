// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph render check Graph svg render with stable data 1`] = `"<svg width=\\"300\\" height=\\"300\\"><path d=\\"M 0 0 L 0 300\\" stroke=\\"grey\\" stroke-width=\\"2\\" fill=\\"none\\"></path><path d=\\"M 0 300 L 300 300\\" stroke=\\"grey\\" stroke-width=\\"2\\" fill=\\"none\\"></path><path d=\\"M 0 0 L 300 0\\" stroke=\\"grey\\" stroke-width=\\"1\\" class=\\"horizontal-lines\\" fill=\\"none\\"></path><path d=\\"M 0 70 L 300 70\\" stroke=\\"grey\\" stroke-width=\\"1\\" class=\\"horizontal-lines\\" fill=\\"none\\"></path><path d=\\"M 0 140 L 300 140\\" stroke=\\"grey\\" stroke-width=\\"1\\" class=\\"horizontal-lines\\" fill=\\"none\\"></path><path d=\\"M 0 210 L 300 210\\" stroke=\\"grey\\" stroke-width=\\"1\\" class=\\"horizontal-lines\\" fill=\\"none\\"></path><path d=\\"M 0 280 L 300 280\\" stroke=\\"grey\\" stroke-width=\\"1\\" class=\\"horizontal-lines\\" fill=\\"none\\"></path><path d=\\"M 0 0 L 0 300\\" stroke=\\"grey\\" stroke-width=\\"1\\" class=\\"vertical-lines\\" fill=\\"none\\"></path><path d=\\"M 70 0 L 70 300\\" stroke=\\"grey\\" stroke-width=\\"1\\" class=\\"vertical-lines\\" fill=\\"none\\"></path><path d=\\"M 140 0 L 140 300\\" stroke=\\"grey\\" stroke-width=\\"1\\" class=\\"vertical-lines\\" fill=\\"none\\"></path><path d=\\"M 210 0 L 210 300\\" stroke=\\"grey\\" stroke-width=\\"1\\" class=\\"vertical-lines\\" fill=\\"none\\"></path><path d=\\"M 280 0 L 280 300\\" stroke=\\"grey\\" stroke-width=\\"1\\" class=\\"vertical-lines\\" fill=\\"none\\"></path><path d=\\"M 1 3 L 1 3 L 2 5 L 3 2 L 4 16 L 18 5\\" stroke=\\"red\\" stroke-width=\\"3\\" class=\\"graphic\\" fill=\\"none\\"></path></svg>"`;
